catch_all generates compilation errors and incorrect discriminant when it is not the last variant

This example…

```rust
use num_enum::FromPrimitive;

#[derive(FromPrimitive)]
#[repr(u8)]
pub enum Number {
    Zero = 0,
    #[num_enum(catch_all)]
    Multiple(u8) = 2,
    One = 1,
}
```

produces this compiler error:

```
error[E0308]: `match` arms have incompatible types
 --> src/lib.rs:3:10
  |
3 | #[derive(FromPrimitive)]
  |          ^^^^^^^^^^^^^
  |          |
  |          expected `Number`, found enum constructor
  |          this is found to be of type `Number`
  |          `match` arms have incompatible types
...
8 |     Multiple(u8) = 2,
  |     -------- `Multiple` defines an enum variant constructor here, which should be called
  |
  = note:          expected enum `Number`
          found enum constructor `fn(u8) -> Number {Number::Multiple}`
  = note: this error originates in the derive macro `FromPrimitive` (in Nightly builds, run with -Z macro-backtrace for more info)
```

Running `cargo expand` is illuminating:

```rust
impl ::num_enum::FromPrimitive for Number {
    type Primitive = u8;
    fn from_primitive(number: Self::Primitive) -> Self {
        #![allow(non_upper_case_globals)]
        const Zero__num_enum_0__: u8 = 0;
        // This is the wrong integer value for this variant:
        const One__num_enum_0__: u8 = 2;
        #[deny(unreachable_patterns)]
        match number {
            Zero__num_enum_0__ => Self::Zero,
            // And this is mapping this constant to the wrong variant:
            One__num_enum_0__ => Self::Multiple,
            #[allow(unreachable_patterns)]
            _ => Self::Multiple(number),
        }
    }
}
```

Everything works as expected, however, if the `catch_all` variant is moved to the end:

```rust
use num_enum::FromPrimitive;

#[derive(FromPrimitive)]
#[repr(u8)]
pub enum Number {
    Zero = 0,
    One = 1,
    #[num_enum(catch_all)]
    Multiple(u8) = 2,
}
```

--- a/num_enum/tests/from_primitive.rs
+++ b/num_enum/tests/from_primitive.rs
@@ -114,6 +114,30 @@ fn from_primitive_number_catch_all() {
     assert_eq!(two, Enum::NonZero(2_u8));
 }
 
+#[test]
+fn from_primitive_number_catch_all_in_middle() {
+    #[derive(Debug, PartialEq, Eq, FromPrimitive)]
+    #[repr(u8)]
+    enum Enum {
+        Zero = 0,
+        #[num_enum(catch_all)]
+        Else(u8) = 2,
+        One = 1,
+    }
+
+    let zero = Enum::from_primitive(0_u8);
+    assert_eq!(zero, Enum::Zero);
+
+    let one = Enum::from_primitive(1_u8);
+    assert_eq!(one, Enum::One);
+
+    let two = Enum::from_primitive(2_u8);
+    assert_eq!(two, Enum::Else(2_u8));
+
+    let three = Enum::from_primitive(3_u8);
+    assert_eq!(three, Enum::Else(3_u8));
+}
+
 #[cfg(feature = "complex-expressions")]
 #[test]
 fn from_primitive_number_with_inclusive_range() {

--- a/num_enum_derive/src/parsing.rs
+++ b/num_enum_derive/src/parsing.rs
@@ -61,6 +61,7 @@ impl EnumInfo {
     pub(crate) fn variant_idents(&self) -> Vec<Ident> {
         self.variants
             .iter()
+            .filter(|variant| !variant.is_catch_all)
             .map(|variant| variant.ident.clone())
             .collect()
     }
@@ -81,6 +82,7 @@ impl EnumInfo {
     pub(crate) fn variant_expressions(&self) -> Vec<Vec<Expr>> {
         self.variants
             .iter()
+            .filter(|variant| !variant.is_catch_all)
             .map(|variant| variant.all_values().cloned().collect())
             .collect()
     }

